tools/GitHubTools.psm1

```diff
@@ -1,23 +1,177 @@
+    [string]$Body
+    [string]$ContributorName
+    [string]$ContributorEmail
+    [string[]]$CommitLabels
+    [int]$PRNumber = -1
+}
+
+class GitHubCommitInfo : GitCommitInfo
+{
+    [string]$Organization
+    [string]$Repository
+    [pscustomobject]$GitHubCommitData
+}
+
+class GitHubIssueInfo
+{
+    [int]$Number
+    [string]$Organization
+    [string]$Repository
+
+    [uri]GetHtmlUri()
+    {
+        return [uri]"https://github.com/$($this.Organization)/$($this.Repository)/issues/$($this.Number)"
+    }
+
+    [uri]GetApiUri()
+    {
+        return [uri]"https://api.github.com/repos/$($this.Organization)/$($this.Repository)/issues/$($this.Number)"
+    }
+}
+
+class GitHubIssue : GitHubIssueInfo
+{
+    [pscustomobject]$RawResponse
+    [string]$Body
+    [string[]]$Labels
+}
+
+class GitHubPR : GitHubIssue
+{
+    hidden [GitHubIssueInfo[]]$ClosedIssues = $null
+
+    [GitHubIssueInfo[]]GetClosedIssueInfos()
+    {
+        if ($null -eq $this.ClosedIssues)
+        {
+            $this.ClosedIssues = $this.Body |
+                GetClosedIssueUrisInBodyText |
+                GetGitHubIssueFromUri
+        }
+
+        return $this.ClosedIssues
+    }
+}
+
+$script:CloseKeywords = @(
+    'close'
+    'closes'
+    'closed'
+    'fix'
+    'fixes'
+    'fixed'
+    'resolve'
+    'resolves'
+    'resolved'
+)
+$script:EndNonCharRegex = [regex]::new('[^0-9]*$', 'compiled')
+filter GetClosedIssueUrisInBodyText
 {
     param(
+        [Parameter(ValueFromPipeline)]
         [string]
-        $Repository
+        $Text
+    )
+
+    $words = $Text.Split()
+
+    $expectIssue = $false
+    for ($i = 0; $i -lt $words.Length; $i++)
+    {
+        $currWord = $words[$i]
+
+        if ($script:CloseKeywords -contains $currWord)
+        {
+            $expectIssue = $true
+            continue
+        }
+
+        if (-not $expectIssue)
+        {
+            continue
+        }
+
+        $expectIssue = $false
+
+        $trimmedWord = $script:EndNonCharRegex.Replace($currWord, '')
+
+        if ([uri]::IsWellFormedUriString($trimmedWord, 'Absolute'))
+        {
+            # Yield
+            [uri]$trimmedWord
+        }
+    }
+}
+
+filter GetGitHubIssueFromUri
+{
+    param(
+        [Parameter(ValueFromPipeline)]
+        [uri]
+        $IssueUri
     )
 
-    if ($Repository.EndsWith('.git'))
+    if ($IssueUri.Authority -ne 'github.com')
     {
-        $Repository = $Repository.Substring(0, $Repository.Length - 4)
+        return
+    }
+
+    if ($IssueUri.Segments.Length -ne 5)
+    {
+        return
+    }
+
+    if ($IssueUri.Segments[3] -ne 'issues/')
+    {
+        return
+    }
+
+    $issueNum = -1
+    if (-not [int]::TryParse($IssueUri.Segments[4], [ref]$issueNum))
+    {
+        return
+    }
+
+    return [GitHubIssueInfo]@{
+        Organization = $IssueUri.Segments[1].TrimEnd('/')
+        Repository = $IssueUri.Segments[2].TrimEnd('/')
+        Number = $issueNum
+    }
+}
+
+filter GetHumanishRepositoryDetails
```

Comment:
LOL!

Reply:
`git` seems to like the `ish` suffix to refer to softened parameter "types", like [commitish](https://stackoverflow.com/questions/23303549/what-are-commit-ish-and-tree-ish-in-git) and [humanish](https://stackoverflow.com/questions/13839879/how-to-determine-the-humanish-part-of-a-git-repository)

Reply:
I'm familiar with "commitish" but not "humanish". Thx.

Reply:
Aggravatingly they use the terms in passing and you have to sift through StackOverflow to decode them